### PR TITLE
Async Interface IAfterInteract()

### PR DESCRIPTION
**Pull request by DmitriyRubetskoy**
I re-wrote the interface and implementations of void IAfterInteract to make it a async Task, rather then a void, so it could be compatible with more functions such as DoAfter().

Please refuse the changes to the yml file and the removal of "IAfterInteract" in the components list of the mop, I tried removing those myself and making a separate commit but git doesn't read it as "changed", so the only way is to be cancelled by maintainers, thanks for your attention,

### CODE REVIEW

File: Resources/Prototypes/Entities/Objects/Specific/chemistry.yml [hunk 1]
```diff
@@ -17,7 +17,6 @@
   - type: Pourable
     transferAmount: 5.0
   - type: Spillable
-  - type: GlassBeaker
```
Comment (PaulRitter) [inline Resources/Prototypes/Entities/Objects/Specific/chemistry.yml:20]: grafik

File: Resources/Prototypes/Entities/Objects/Specific/chemistry.yml [hunk 2]
```diff
@@ -38,7 +37,7 @@
...
```
Comment (PaulRitter) [inline Resources/Prototypes/Entities/Objects/Specific/chemistry.yml:41]: grafik

File: Content.Server/GameObjects/Components/Fluids/MopComponent.cs [hunk 3]
```diff
@@ -9,14 +9,16 @@
...
+using System.Threading.Tasks;
 
 namespace Content.Server.GameObjects.Components.Fluids
 {
     /// <summary>
     /// For cleaning up puddles
     /// </summary>
     [RegisterComponent]
-    public class MopComponent : Component, IAfterInteract
+    public class MopComponent : Component,
```
Comment (PaulRitter) [inline Content.Server/GameObjects/Components/Fluids/MopComponent.cs:20]:
```suggestion
    public class MopComponent : Component, IAfterInteract
```

**Review by PaulRitter — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Entities/Objects/Specific/chemistry.yml (hunk 2 above)
Comment (DmitriyRubetskoy) [reply Resources/Prototypes/Entities/Objects/Specific/chemistry.yml:41]: I have zero idea why the beaker decided not to be a beaker anymore

File: Resources/Prototypes/Entities/Objects/Specific/chemistry.yml (hunk 1 above)
Comment (DmitriyRubetskoy) [reply Resources/Prototypes/Entities/Objects/Specific/chemistry.yml:20]: can you please mark this as a suggested change again? It doesn't give me an option to commit the suggestion for some reason.

File: Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs [hunk 4]
```diff
@@ -15,7 +16,7 @@ public interface IAfterInteract
         /// <summary>
         /// Called when we interact with nothing, or when we interact with an entity out of range that has no behavior
         /// </summary>
-        void AfterInteract(AfterInteractEventArgs eventArgs);
+        Task IAfterInteract(AfterInteractEventArgs eventArgs);
```
Comment (PaulRitter) [inline Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs:19]:
```suggestion
        async Task IAfterInteract(AfterInteractEventArgs eventArgs);
```

**Review by PaulRitter — CHANGES_REQUESTED** [review]

**Comment by PaulRitter** [comment]
lmao im retarded revert that latest review

File: Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs [hunk 5]
```diff
@@ -15,7 +16,7 @@ public interface IAfterInteract
         /// <summary>
         /// Called when we interact with nothing, or when we interact with an entity out of range that has no behavior
         /// </summary>
-        void AfterInteract(AfterInteractEventArgs eventArgs);
+        async Task IAfterInteract(AfterInteractEventArgs eventArgs);
```
Comment (PaulRitter) [inline Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs:19]:
```suggestion
        Task IAfterInteract(AfterInteractEventArgs eventArgs);
```

**Review by PJB3005 — CHANGES_REQUESTED** [review]
You didn't modify the code in `InteractionSystem` so the tasks are not consumed correctly there.

**Comment by DmitriyRubetskoy** [comment]
> You didn't modify the code in `InteractionSystem` so the tasks are not consumed correctly there.

Fixed, please review if results are properly coded

**Comment by PJB3005** [comment]
Did you forget to push? You didn't make any changes to `InteractionSystem`.

**Comment by DmitriyRubetskoy** [comment]
Yes, something went wrong with the push. Changes are now commited.

**Comment by metalgearsloth** [comment]
Does this build locally? github is saying it dun build

File: Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs (hunk 4 above)
Comment (PJB3005) [inline Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs:19]: You probably didn't intend to rename this method?

Comment (DmitriyRubetskoy) [reply Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs:19]: I am not sure what the right call was there, I followed the refactoring tips.
It is a separate method I mistakenly renamed, right?

Comment (DmitriyRubetskoy) [reply Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs:19]: Although the method was originally named IAfterInteract, wasn't it?

Comment (PaulRitter) [reply Content.Shared/Interfaces/GameObjects/Components/Interaction/IAfterInteract.cs:19]: no, the method was originially called `AfterInteract`. The **Interface** it is part of is called `IAfterInteract`.

**Comment by SweptWasTaken** [comment]
You should honestly go take a crash course on git or get a decent git gui, you've made a couple PRs here now that are completely borked due to you committing seemingly random stuff...

**Comment by DmitriyRubetskoy** [comment]
> You should honestly go take a crash course on git or get a decent git gui, you've made a couple PRs here now that are completely borked due to you committing seemingly random stuff...

Reasonable, some of the things which get into my commits I myself wonder how did they work out in such a way, but for right now I'd like to first finish this piece of work so I can go work on something simpler, which in turn relies on this being implemented.
Is there anything else need fixing at the moment?

**Comment by DmitriyRubetskoy** [comment]
Are there any parts which need alteration or is something so deeply wrong it is better to just close the pull request?

**Comment by metalgearsloth** [comment]
> 
> 
> Are there any parts which need alteration or is something so deeply wrong it is better to just close the pull request?

It still doesn't build so you'll need to fix that before it can be reviewed https://github.com/space-wizards/space-station-14/pull/2735/checks?check_run_id=1539144916

**Comment by DmitriyRubetskoy** [comment]
Understood. Will be working on it during this week.

**Comment by DmitriyRubetskoy** [comment]
Checks run good. Merge?

File: Resources/Prototypes/Entities/Objects/Specific/chemistry.yml [hunk 6]
```diff
@@ -39,6 +39,7 @@
     transferAmount: 5.0
   - type: Spillable
   - type: GlassBeaker
+
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Objects/Specific/chemistry.yml:42]:
```suggestion
```

File: Content.Server/GameObjects/Components/Portal/TeleporterComponent.cs [hunk 7]
```diff
@@ -22,7 +23,8 @@
 namespace Content.Server.GameObjects.Components.Portal
 {
     [RegisterComponent]
-    public class TeleporterComponent : Component, IAfterInteract
+    public class TeleporterComponent : Component, IAfterInteract 
+
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Portal/TeleporterComponent.cs:27]:
```suggestion
```

File: Content.Server/GameObjects/Components/ActionBlocking/HandcuffComponent.cs [hunk 8]
```diff
@@ -21,6 +22,7 @@ namespace Content.Server.GameObjects.Components.ActionBlocking
 {
     [RegisterComponent]
     public class HandcuffComponent : SharedHandcuffComponent, IAfterInteract
+
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/ActionBlocking/HandcuffComponent.cs:25]:
```suggestion
```

File: Content.Server/GameObjects/Components/Chemistry/InjectorComponent.cs [hunk 9]
```diff
@@ -106,7 +107,7 @@ private void Toggle(IEntity user)
         /// Called when clicking on entities while holding in active hand
         /// </summary>
         /// <param name="eventArgs"></param>
-        void IAfterInteract.AfterInteract(AfterInteractEventArgs eventArgs)
+        public async Task AfterInteract(AfterInteractEventArgs eventArgs)
         {
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/InjectorComponent.cs:111]: Kinda weird to change these to public

File: Content.Server/GameObjects/Components/Fluids/RagComponent.cs [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+using Robust.Shared.Serialization;
+using System.Threading.Tasks;
+
+namespace Content.Server.GameObjects.Components.Fluids
+{
+    /// <summary>
+    /// For cleaning up puddles
+    /// </summary>
+    [RegisterComponent]
+    public class RagComponent : Component, IAfterInteract
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Fluids/RagComponent.cs:20]: Is rag supposed to be in this PR?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/GameObjects/Components/Chemistry/InjectorComponent.cs (hunk 9 above)
Comment (DmitriyRubetskoy) [reply Content.Server/GameObjects/Components/Chemistry/InjectorComponent.cs:111]: I re-compiled the code several times and fixed the issues which croped up with certain methods being public, this one wasn't one of them.
Maybe the right thing to do is to make it just async but it didn't cause any issues

File: Content.Server/GameObjects/Components/Fluids/RagComponent.cs (hunk 10 above)
Comment (DmitriyRubetskoy) [reply Content.Server/GameObjects/Components/Fluids/RagComponent.cs:20]: No actually, I believe it got in there by mistake. I'll commit the changes.

Comment (metalgearsloth) [reply Content.Server/GameObjects/Components/Fluids/RagComponent.cs:20]: I'm happy to merge if it's removed but I won't be able to do it for a bit.

**Comment by DmitriyRubetskoy** [comment]
All done.

**Comment by metalgearsloth** [comment]
> 
> 
> All done.
> 
> EDIT: You were wrong, the Injector has to be a public function, it causes errors otherwise.

I can make mistakes but in this case I'm right.

Source: Half of these are still explicit implementations.

**Comment by DmitriyRubetskoy** [comment]
The function does not compile if not stated as a public async and causes a build error.

**Comment by metalgearsloth** [comment]
You seem to be confused because I'm talking about the ones you changed from explicit interface implementations to just being public methods like the one I commented on.

File: Content.Server/GameObjects/Components/Fluids/SprayComponent.cs [hunk 11]
```diff
@@ -98,7 +99,7 @@ public override void ExposeData(ObjectSerializer serializer)
             serializer.DataField(ref _safety, "safety", true);
         }
 
-        void IAfterInteract.AfterInteract(AfterInteractEventArgs eventArgs)
+         async Task IAfterInteract.AfterInteract(AfterInteractEventArgs eventArgs)
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Fluids/SprayComponent.cs:102]:
```suggestion
        async Task IAfterInteract.AfterInteract(AfterInteractEventArgs eventArgs)
```

File: Content.Server/GameObjects/Components/Fluids/MopComponent.cs [hunk 12]
```diff
@@ -17,6 +18,7 @@ namespace Content.Server.GameObjects.Components.Fluids
     /// </summary>
     [RegisterComponent]
     public class MopComponent : Component, IAfterInteract
+
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Fluids/MopComponent.cs:21]:
```suggestion
```

**Review by metalgearsloth — APPROVED** [review]

**Comment by DmitriyRubetskoy** [comment]
Are there more explicit implementations or other then the injector everything else seems in order?

**Pull request merged**